libraries/heft-config-file/src/ConfigurationFile.ts

```diff
@@ -199,187 +228,214 @@ export class ConfigurationFile<TConfigurationFile> {
 
   private async _loadConfigurationFileInnerAsync(
     resolvedConfigurationFilePath: string,
+    options: ILoadConfigurationFileOptions,
+    isFirstFile: boolean,
     visitedConfigurationFilePaths: Set<string>
-  ): Promise<TConfigurationFile> {
+  ): Promise<TConfigurationFile | undefined> {
     let cacheEntry:
       | IConfigurationFileCacheEntry<TConfigurationFile>
       | undefined = this._configurationFileCache.get(resolvedConfigurationFilePath);
     if (!cacheEntry) {
-      try {
-        const resolvedConfigurationFilePathForErrors: string = ConfigurationFile._formatPathForError(
-          resolvedConfigurationFilePath
-        );
-
-        if (visitedConfigurationFilePaths.has(resolvedConfigurationFilePath)) {
-          throw new Error(
-            'A loop has been detected in the "extends" properties of configuration file at ' +
-              `"${resolvedConfigurationFilePathForErrors}".`
-          );
-        }
+      cacheEntry = await this._generateCacheEntryAsync(
+        resolvedConfigurationFilePath,
+        options,
+        visitedConfigurationFilePaths
+      );
+      this._configurationFileCache.set(resolvedConfigurationFilePath, cacheEntry);
+    }
 
-        visitedConfigurationFilePaths.add(resolvedConfigurationFilePath);
+    if (cacheEntry.parentFileError) {
+      throw cacheEntry.parentFileError;
+    } else if (cacheEntry.error) {
+      if (!isFirstFile || !options.ignoreIfNotExist || !FileSystem.isNotExistError(cacheEntry.error)) {
```

Comment:
@iclanton This is maybe offtopic, but right now when Heft **_successfully_** compiles a small project, it throws and catches 14 NOENT errors, almost half of which involve config files:

```
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\build-cache\lib\index.d.ts
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\build-cache\lib\index.d.ts.map
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\build-cache\lib\index.js
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\build-cache\lib\index.js.map
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\build-cache\lib\test\ExampleTest.test.d.ts
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\build-cache\lib\test\ExampleTest.test.d.ts.map
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\build-cache\lib\test\ExampleTest.test.js
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\build-cache\lib\test\ExampleTest.test.js.map
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\copy-static-assets.json
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\copy-static-assets.json       
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\plugins.json       
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\plugins.json       
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\typescript.json       
File does not exist: C:\Git\rushstack2\build-tests\heft-node-basic-test\.heft\typescript.json       
```

This seems inefficient. It also destroys our ability to use the "break on exception" feature of the debugger to trap real problems.

I feel like `FileSystem.exists()` would be an intelligent way to solve these problems.  

It introduces a theoretical race condition of a file getting deleted immediately after `FileSystem.exists()` returns true, but that seems like not a realistic concern for the actual problem we're solving here.

Reply:
It reminds me of the old C# antipattern:

```ts
string value;
try {
  value = dictionary[key];
} 
catch (KeyNotFoundException exception) {
  value = null;
}
```
...which ultimately was deprecated in favor of the `Dictionary.tryGetValue()` API.